### PR TITLE
JAMES-2811 BlobStore Vault retention implementation

### PR DESCRIPTION
**Pull request by trantienduchn**
- Need to take care about the webadmin retention [not done]
- BlobStoreDeletedMessagesVault need to implement `DeletedMessageVaultSearchContract.AllContracts`

### CODE REVIEW

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BucketNameGenerator.java [hunk 1]
```diff
@@ -20,24 +20,62 @@
...
         ZonedDateTime now = ZonedDateTime.now(clock);
-        int month = now.getMonthValue();
+        int month = now.getMonth().get(ChronoField.MONTH_OF_YEAR);
         int year = now.getYear();
-        return BucketName.of(String.format("deleted-messages-%d-%02d-01", year, month));
+        return BucketName.of(String.format(BUCKET_NAME_GENRATING_FORMAT, year, month));
+    }
+
+    @VisibleForTesting
+    boolean isBucketRangeBefore(BucketName bucketName, ZonedDateTime beforeThisTime) {
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BucketNameGenerator.java]: IMO you can simply return the ZoneDateTime that this bucket started being used at, and let the caller do the comparison (beforeThisTime parameter should be removed IMO)

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BucketNameGenerator.java [hunk 2]
```diff
@@ -20,24 +20,62 @@
...
+        return bucketStartingTime(bucketName)
+            .map(bucketStartingTime -> bucketStartingTime.isBefore(firstTimeOfTheMonth(beforeThisTime)))
+            .orElse(false);
+    }
+
+    private ZonedDateTime firstTimeOfTheMonth(ZonedDateTime moment) {
+        return moment.withDayOfMonth(1).with(LocalTime.MIN);
+    }
+
+    private Optional<ZonedDateTime> bucketStartingTime(BucketName bucketName) {
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BucketNameGenerator.java]: Let's keep that as a public method

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java [hunk 3]
```diff
@@ -96,6 +106,23 @@
...
+        Flux<BucketName> deleteOperation = retentionQualifiedBuckets(beginningOfRetentionPeriod)
+            .flatMap(bucketName -> deleteBucketData(bucketName).then(Mono.just(bucketName)));
+
+        return new BlobStoreVaultGarbageCollectionTask(beginningOfRetentionPeriod, deleteOperation);
+    }
+
+    @VisibleForTesting
+    Flux<BucketName> retentionQualifiedBuckets(ZonedDateTime beginningOfRetentionPeriod) {
+        return Flux.from(messageMetadataVault.listRelatedBuckets())
+            .filter(bucketName -> nameGenerator.isBucketRangeBefore(bucketName, beginningOfRetentionPeriod));
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java]: In bucket `dmsg-2018-07-01` we have email ranging from 01/07/2018 to 31/07/2018.

We are 10/07/2018. I apply retention. You are using the starting date for the 1 year retention. Thus I will delete mails of the 23/07/2019. Which is not already 1 year.

We should use the ending date associated with the bucket, not the starting one. Carefull about month being 28-29-30-31 days when doing the maths...

Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java]: I would prefer a 

```
buckets.filter(bucket -> generator.endDate(bucket).isBefore(retentionStart));
```

Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java]: What happens when the name pattern does not match? (we should ERROR log + not fail)

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                .map(BucketName::asString)
+                .collect(Guavate.toImmutableList());
+        }
+    }
+
+    public static final String TYPE = "deletedMessages/blobStoreBasedGarbageCollection";
+
+    private final Flux<BucketName> retentionOperation;
+    private final ZonedDateTime beginningOfRetentionPeriod;
+    private final Vector<BucketName> deletedBuckets;
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java]: Why vector? Don't we have better mutable datastructure?

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public Result run() {
+        retentionOperation
+            .doOnNext(deletedBuckets::add)
+            .subscribeOn(Schedulers.elastic())
+            .then()
+            .block();
+
+        return Result.COMPLETED;
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java]: What happen if a bucket cannot be deleted? We Throw?

File: mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVaultTest.java [hunk 6]
```diff
@@ -19,110 +19,112 @@
...
+    @Test
+    void retentionQualifiedBucketsShouldReturnOnlyBucketsFullyBeforeBeginningOfRetentionPeriod() {
+        clock.setInstant(Instant.parse("2007-12-03T10:15:30.00Z"));
+        Mono.from(getVault().append(USER, OLD_DELETED_MESSAGE, new ByteArrayInputStream(CONTENT))).block();
+        clock.setInstant(Instant.parse("2008-01-03T10:15:30.00Z"));
+        Mono.from(getVault().append(USER, DELETED_MESSAGE_2, new ByteArrayInputStream(CONTENT))).block();
 
+        ZonedDateTime beginningOfRetention = ZonedDateTime.parse("2008-01-30T10:15:30.00Z");
+        assertThat(messageVault.retentionQualifiedBuckets(beginningOfRetention).toStream())
+            .containsOnly(BucketName.of("deleted-messages-2007-12-01"));
```
Comment (chibenwa) [inline mailbox/plugin/deleted-messages-vault/src/test/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVaultTest.java]: This bucket also contains mail from `2007/12/25` that are not yet 1 year old and that you would delete. CF earlier comments.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 7]
```diff
@@ -606,6 +608,25 @@ public void vaultPurgeShouldMakeExportProduceEmptyZipWhenAllMessagesAreExpired()
...
+        purgeVault(webAdminApi);
+
+        String fileLocation = exportAndGetFileLocationFromLastMail(EXPORT_ALL_HOMER_MESSAGES_TO_BART, bartAccessToken);
+        try (ZipAssert zipAssert = assertThatZip(new FileInputStream(fileLocation))) {
+            zipAssert.hasNoEntry();
+        }
+    }
+
+    @Test
+    public void vaultPurgeShouldMakeExportProduceEmptyZipWhenAllMessagesAreExpiredJustOneDay() throws Exception {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: IMO we don't need to test that corner test

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java (hunk 4 above)
Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java]: I want to have a mutable list to be able to have real time update for the task detail. If I use immutable list, the moment I can able to collect all of deleted buckets is only the moment the task finish

Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java]: How about arrayList?

Comment (trantienduchn) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java]: Vector is thread safe and ArrayList is not

**Comment by chibenwa** [comment]
We should split this PR in two so that it do not depend on Integration tests

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java (hunk 4 above)
Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/utils/BlobStoreVaultGarbageCollectionTask.java]: ConcurrentDeque then we donnot care about ordering

**Comment by chibenwa** [comment]
https://github.com/linagora/james-project/pull/2519 is a subpart of this work, contributing retention without dependencies

**Comment by trantienduchn** [comment]
the impl is done in another PR

**Pull request closed without merging**